From df21c5131895af8b1bf21f822d9244e6789b191a Mon Sep 17 00:00:00 2001
From: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
Date: Thu, 3 Feb 2022 11:01:33 +0100
Subject: [PATCH 075/163] Fix infinite loop in dash stroker

When the positions were large enough, we would get precision errors
with the floating point numbers.

We calculate the relative position:
  dpos = pos + dashes[istart] - doffset - estart
and then later
  pos = dpos + estart

If estart is a huge number (range of 10^18) and dashes[istart] is
a low number (10^1), then estart + dashes[istart] == estart and
the loop will never progress, dpos will be 0.

Since the loop should never iterate more than dashCount times
(since we cut away all full dash sequences before entering it),
we add a failsafe that exits the loop if it detects the error.

Fixes: QTBUG-99690
Pick-to: 5.15 6.2 6.3
Change-Id: Ia6a42f21b6b4c6adf5cdd703b6483750513a88ba
Reviewed-by: Lars Knoll <lars.knoll@qt.io>
(cherry picked from commit 26a0638222933fc549f250333c28d0184205b704)

Changed a bit by Albert Astals Cid so it applies to the kde/5.15 branch
---
 src/gui/painting/qstroker.cpp | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/gui/painting/qstroker.cpp b/src/gui/painting/qstroker.cpp
index 22302f9790..07fd1ac105 100644
--- a/src/gui/painting/qstroker.cpp
+++ b/src/gui/painting/qstroker.cpp
@@ -1179,6 +1179,7 @@ void QDashStroker::processCurrentSubpath()
 
         bool done = pos >= estop;
 
+        int maxDashes = dashCount;
         if (clipping) {
             // Check if the entire line can be clipped away.
             if (!lineIntersectsRect(prev, e, clip_tl, clip_br)) {
@@ -1196,7 +1197,7 @@ void QDashStroker::processCurrentSubpath()
                         done = true;
                     } else { // Dash is on this line
                         pos = dpos + estart;
-                        done = pos >= estop;
+                        done = --maxDashes > 0 ? pos >= estop : estop;
                         if (++idash >= dashCount)
                             idash = 0;
                         doffset = 0; // full segment so no offset on next.
-- 
2.37.2

